### PR TITLE
feat: add C `ndarray` implementation for `blas/base/dscal`

### PR DESCRIPTION
**Pull request by aman-095**
Progresses #2039.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C `ndarray` API for `blas/base/dscal`.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js [hunk 1]
```diff
@@ -43,39 +43,10 @@ var M = 5;
...
-			x[ i+4 ] *= alpha;
-		}
-		return x;
-	}
-	N *= stride;
-	for ( i = 0; i < N; i += stride ) {
-		x[ i ] *= alpha;
-	}
-	return x;
+	return ndarray( N, alpha, x, stride, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js]: @aman-095 I believe we want to support negative strides here. That means using `stride2offset` to compute the offset before calling into the `ndarray` API.

File: lib/node_modules/@stdlib/blas/base/dscal/src/dscal.c [hunk 2]
```diff
@@ -28,36 +28,8 @@
...
-			X[ i+3 ] *= alpha;
-			X[ i+4 ] *= alpha;
-		}
-		return;
-	}
-	for ( i = 0; i < N*stride; i += stride ) {
-		X[ i ] *= alpha;
-	}
-	return;
+	API_SUFFIX(c_dscal_ndarray)( N, alpha, X, stride, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/dscal/src/dscal.c]: @aman-095 Similarly, we should go ahead and support negative strides here.

File: lib/node_modules/@stdlib/blas/base/dscal/src/dscal_cblas.c [hunk 3]
```diff
@@ -31,3 +32,17 @@
...
+* Multiplies a double-precision floating-point vector `X` by a constant using alternative indexing semantics.
+*
+* @param N       number of indexed elements
+* @param alpha   scalar
+* @param X       input array
+* @param stride  index increment
+* @param offset  starting index
+*/
+void API_SUFFIX(c_dscal_ndarray)( const CBLAS_INT N, const double alpha, double *X, const CBLAS_INT stride, const CBLAS_INT offset ) {
+	X += stdlib_strided_min_view_buffer_index( N, stride, offset ); // adjust array pointer
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/dscal/src/dscal_cblas.c:51]: @aman-095 Before L47, we should convert `stride` to a positive stride before calling into CBLAS, which may or may not support negative strides. By always passing in a positive stride, we avoid backward compat breaks.

File: lib/node_modules/@stdlib/blas/base/dscal/src/dscal_cblas.c [hunk 4]
```diff
@@ -31,3 +32,17 @@
 void API_SUFFIX(c_dscal)( const CBLAS_INT N, const double alpha, double *X, const CBLAS_INT stride ) {
 	API_SUFFIX(cblas_dscal)( N, alpha, X, stride );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/dscal/src/dscal_cblas.c]: @aman-095 We should also support negative strides here.

File: lib/node_modules/@stdlib/blas/base/dscal/src/dscal_f.c [hunk 5]
```diff
@@ -31,3 +32,21 @@
 void API_SUFFIX(c_dscal)( const CBLAS_INT N, const double alpha, double *X, const CBLAS_INT stride ) {
 	dscal( &N, &alpha, X, &stride );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/dscal/src/dscal_f.c]: @aman-095 We should also support negative strides here.

File: lib/node_modules/@stdlib/blas/base/dscal/src/dscal_f.c [hunk 6]
```diff
@@ -31,3 +32,21 @@
...
+*
+* @param N       number of indexed elements
+* @param alpha   scalar
+* @param X       input array
+* @param stride  index increment
+* @param offset  starting index
+*/
+void API_SUFFIX(c_dscal_ndarray)( const CBLAS_INT N, const double alpha, double *X, const CBLAS_INT stride, const CBLAS_INT offset ) {
+	CBLAS_INT sx = stride;
+	if( sx < 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/dscal/src/dscal_f.c]:
```suggestion
	if ( sx < 0 ) {
```

File: lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js (hunk 1 above)
Comment (aman-095) [reply lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js]: But, native `dscal` implementation from `lapack` doesn't allow for negative stride.

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js]: That was my point about converting from negative to positive. But here, our `ndarray.js` file should support negative strides. In C, we should convert the stride before calling CBLAS. Meaning, *we* are free to support negative strides, but we need to ensure that if and when we call a library following LAPACK that we only pass along positive strides. In short, for all our Level 1 routines, we should be consistent in supporting negative strides. The current situation of some supporting and others not supporting is just confusing.

Comment (aman-095) [reply lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js]: It means that for our Js implementation, we should allow negative strides for all BLAS Level 1 and only for `CBLAS`, or other hardware-optimized libraries we use the same convention of stride conversion.

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/base/dscal/lib/dscal.js]: Correct. We need to update our JS implementations to support negative strides.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @aman-095!

**Pull request merged**